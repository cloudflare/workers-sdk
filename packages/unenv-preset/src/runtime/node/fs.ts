// Hybrid polyfill for `node:fs`
// Needed as a few symbols are missing from the native implementation

// Missing from workerd native `fs` implementation
import {
	F_OK,
	FileReadStream,
	FileWriteStream,
	openAsBlob,
	R_OK,
	W_OK,
	X_OK,
} from "unenv/node/fs";
import type nodeFs from "node:fs";

// Export unenv symbols
export {
	FileReadStream,
	FileWriteStream,
	openAsBlob,
	F_OK,
	R_OK,
	W_OK,
	X_OK,
} from "unenv/node/fs";

const workerdFs = process.getBuiltinModule("node:fs");

const {
	access,
	accessSync,
	appendFile,
	appendFileSync,
	chmod,
	chmodSync,
	chown,
	chownSync,
	close,
	closeSync,
	constants,
	copyFile,
	copyFileSync,
	cp,
	cpSync,
	createReadStream,
	createWriteStream,
	Dir,
	Dirent,
	exists,
	existsSync,
	fchmod,
	fchmodSync,
	fchown,
	fchownSync,
	fdatasync,
	fdatasyncSync,
	fstat,
	fstatSync,
	fsync,
	fsyncSync,
	ftruncate,
	ftruncateSync,
	futimes,
	futimesSync,
	glob,
	globSync,
	lchmod,
	lchmodSync,
	lchown,
	lchownSync,
	link,
	linkSync,
	lstat,
	lstatSync,
	lutimes,
	lutimesSync,
	mkdir,
	mkdirSync,
	mkdtemp,
	mkdtempSync,
	open,
	opendir,
	opendirSync,
	openSync,
	promises,
	read,
	readdir,
	readdirSync,
	readFile,
	readFileSync,
	readlink,
	readlinkSync,
	ReadStream,
	readSync,
	readv,
	readvSync,
	realpath,
	realpathSync,
	rename,
	renameSync,
	rm,
	rmdir,
	rmdirSync,
	rmSync,
	stat,
	statfs,
	statfsSync,
	Stats,
	statSync,
	symlink,
	symlinkSync,
	truncate,
	truncateSync,
	unlink,
	unlinkSync,
	unwatchFile,
	utimes,
	utimesSync,
	watch,
	watchFile,
	write,
	writeFile,
	writeFileSync,
	WriteStream,
	writeSync,
	writev,
	writevSync,
} = workerdFs;

// Export native symbols
export {
	access,
	accessSync,
	appendFile,
	appendFileSync,
	chmod,
	chmodSync,
	chown,
	chownSync,
	close,
	closeSync,
	constants,
	copyFile,
	copyFileSync,
	cp,
	cpSync,
	createReadStream,
	createWriteStream,
	Dir,
	Dirent,
	exists,
	existsSync,
	fchmod,
	fchmodSync,
	fchown,
	fchownSync,
	fdatasync,
	fdatasyncSync,
	fstat,
	fstatSync,
	fsync,
	fsyncSync,
	ftruncate,
	ftruncateSync,
	futimes,
	futimesSync,
	glob,
	globSync,
	lchmod,
	lchmodSync,
	lchown,
	lchownSync,
	link,
	linkSync,
	lstat,
	lstatSync,
	lutimes,
	lutimesSync,
	mkdir,
	mkdirSync,
	mkdtemp,
	mkdtempSync,
	open,
	opendir,
	opendirSync,
	openSync,
	promises,
	read,
	readdir,
	readdirSync,
	readFile,
	readFileSync,
	readlink,
	readlinkSync,
	ReadStream,
	readSync,
	readv,
	readvSync,
	realpath,
	realpathSync,
	rename,
	renameSync,
	rm,
	rmdir,
	rmdirSync,
	rmSync,
	stat,
	statfs,
	statfsSync,
	Stats,
	statSync,
	symlink,
	symlinkSync,
	truncate,
	truncateSync,
	unlink,
	unlinkSync,
	unwatchFile,
	utimes,
	utimesSync,
	watch,
	watchFile,
	write,
	writeFile,
	writeFileSync,
	WriteStream,
	writeSync,
	writev,
	writevSync,
};

export default {
	// unenv
	// @ts-expect-error Not in Node types
	FileReadStream,
	FileWriteStream,
	openAsBlob,
	F_OK,
	R_OK,
	W_OK,
	X_OK,

	// workerd
	access,
	accessSync,
	appendFile,
	appendFileSync,
	chmod,
	chmodSync,
	chown,
	chownSync,
	close,
	closeSync,
	constants,
	copyFile,
	copyFileSync,
	cp,
	cpSync,
	createReadStream,
	createWriteStream,
	Dir,
	Dirent,
	exists,
	existsSync,
	fchmod,
	fchmodSync,
	fchown,
	fchownSync,
	fdatasync,
	fdatasyncSync,
	fstat,
	fstatSync,
	fsync,
	fsyncSync,
	ftruncate,
	ftruncateSync,
	futimes,
	futimesSync,
	glob,
	globSync,
	lchmod,
	lchmodSync,
	lchown,
	lchownSync,
	link,
	linkSync,
	lstat,
	lstatSync,
	lutimes,
	lutimesSync,
	mkdir,
	mkdirSync,
	mkdtemp,
	mkdtempSync,
	open,
	opendir,
	opendirSync,
	openSync,
	promises,
	read,
	readdir,
	readdirSync,
	readFile,
	readFileSync,
	readlink,
	readlinkSync,
	ReadStream,
	readSync,
	readv,
	readvSync,
	realpath,
	realpathSync,
	rename,
	renameSync,
	rm,
	rmdir,
	rmdirSync,
	rmSync,
	stat,
	statfs,
	statfsSync,
	Stats,
	statSync,
	symlink,
	symlinkSync,
	truncate,
	truncateSync,
	unlink,
	unlinkSync,
	unwatchFile,
	utimes,
	utimesSync,
	watch,
	watchFile,
	write,
	writeFile,
	writeFileSync,
	WriteStream,
	writeSync,
	writev,
	writevSync,
} satisfies typeof nodeFs;
