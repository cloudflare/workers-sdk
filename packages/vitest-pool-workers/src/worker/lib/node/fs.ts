function throwNotFound(path: string): never {
	const error = new Error(`ENOENT: no such file or directory, open '${path}'`);
	Object.assign(error, {
		errno: -2,
		syscall: "open",
		code: "ENOENT",
		path,
	});
	throw error;
}

export function existsSync(_path: string) {
	// TODO(soon): figure out what's calling this and why, probably needed for
	//  snapshots
	// console.log("fs.ts:existsSync", path);
	return false;
}

export function readdirSync(path: string) {
	console.log("fs.ts:readdirSync", path);
	return [];
}

export function readFileSync(path: string) {
	console.log("fs.ts:readFileSync", path);
	throwNotFound(path);
}

export class Stats {}
export function statSync(path: string) {
	console.log("fs.ts:readdirSync", path);
	throwNotFound(path);
}

export function realpathSync(path: string) {
	console.log("fs.ts:realpathSync", path);
	throwNotFound(path);
}

export const promises = {
	async readFile(path: string) {
		console.log("fs.ts:promises.readFile", path);
		throwNotFound(path);
	},
};

export default {};
