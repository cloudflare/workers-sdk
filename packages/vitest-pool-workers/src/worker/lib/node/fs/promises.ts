export async function readFile(path: string) {
	console.log("fs/promises.ts:readFile", path);
	const error = new Error(`ENOENT: no such file or directory, open '${path}'`);
	Object.assign(error, {
		errno: -2,
		syscall: "open",
		code: "ENOENT",
		path,
	});
	throw error;
}

export default {
	readFile,
};
