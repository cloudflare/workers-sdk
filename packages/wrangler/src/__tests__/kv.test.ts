import { setMock, unsetAllMocks } from "./mock-cfetch";
import { runWrangler } from "./run-wrangler";
import { runInTempDir } from "./run-in-tmp";
import { writeFileSync } from "fs";

describe("wrangler", () => {
  runInTempDir();

  afterEach(() => {
    unsetAllMocks();
  });

  describe("kv:namespace", () => {
    describe("create", () => {
      function mockCreateRequest(expectedTitle: string) {
        setMock(
          "/accounts/:accountId/storage/kv/namespaces",
          "POST",
          ([_url, accountId], { body }) => {
            expect(accountId).toEqual("some-account-id");
            const title = JSON.parse(body).title;
            expect(title).toEqual(expectedTitle);
            return { id: "some-namespace-id" };
          }
        );
      }

      it("should error if no namespace is given", async () => {
        const { stdout, stderr } = await runWrangler("kv:namespace create");
        expect(stdout).toMatchInlineSnapshot(`""`);
        expect(stderr).toMatchInlineSnapshot(`
          "wrangler kv:namespace create <namespace>

          Create a new namespace

          Positionals:
            namespace  The name of the new namespace  [string] [required]

          Flags:
            -c, --config   Path to .toml configuration file  [string]
            -h, --help     Show help  [boolean]
            -v, --version  Show version number  [boolean]

          Options:
            -l, --local    Run on my machine  [boolean] [default: false]
                --env      Perform on a specific environment  [string]
                --preview  Interact with a preview namespace  [boolean]
          Not enough non-option arguments: got 0, need at least 1"
        `);
      });

      it("should error if the namespace to create contains spaces", async () => {
        const { stdout, stderr } = await runWrangler(
          "kv:namespace create abc def ghi"
        );
        expect(stdout).toMatchInlineSnapshot(`""`);
        expect(stderr).toMatchInlineSnapshot(`
          "wrangler kv:namespace create <namespace>

          Create a new namespace

          Positionals:
            namespace  The name of the new namespace  [string] [required]

          Flags:
            -c, --config   Path to .toml configuration file  [string]
            -h, --help     Show help  [boolean]
            -v, --version  Show version number  [boolean]

          Options:
            -l, --local    Run on my machine  [boolean] [default: false]
                --env      Perform on a specific environment  [string]
                --preview  Interact with a preview namespace  [boolean]
          Unexpected additional positional arguments \\"def ghi\\"."
        `);
      });

      it("should error if the namespace to create is not valid", async () => {
        const { stdout, stderr } = await runWrangler(
          "kv:namespace create abc-def"
        );
        expect(stdout).toMatchInlineSnapshot(`""`);
        expect(stderr).toMatchInlineSnapshot(`
          "wrangler kv:namespace create <namespace>

          Create a new namespace

          Positionals:
            namespace  The name of the new namespace  [string] [required]

          Flags:
            -c, --config   Path to .toml configuration file  [string]
            -h, --help     Show help  [boolean]
            -v, --version  Show version number  [boolean]

          Options:
            -l, --local    Run on my machine  [boolean] [default: false]
                --env      Perform on a specific environment  [string]
                --preview  Interact with a preview namespace  [boolean]
          The namespace binding name \\"abc-def\\" is invalid. It can only have alphanumeric and _ characters, and cannot begin with a number."
        `);
      });

      it("should create a namespace", async () => {
        mockCreateRequest("worker-UnitTestNamespace");
        const { stdout } = await runWrangler(
          "kv:namespace create UnitTestNamespace"
        );
        expect(stdout).toMatchInlineSnapshot(`
          "ðŸŒ€ Creating namespace with title \\"worker-UnitTestNamespace\\"
          âœ¨ Success!
          Add the following to your configuration file in your kv_namespaces array:
          { binding = \\"UnitTestNamespace\\", id = \\"some-namespace-id\\" }"
        `);
      });

      it("should create a preview namespace if configured to do so", async () => {
        mockCreateRequest("worker-UnitTestNamespace_preview");
        const { stdout } = await runWrangler(
          "kv:namespace create UnitTestNamespace --preview"
        );
        expect(stdout).toMatchInlineSnapshot(`
          "ðŸŒ€ Creating namespace with title \\"worker-UnitTestNamespace_preview\\"
          âœ¨ Success!
          Add the following to your configuration file in your kv_namespaces array:
          { binding = \\"UnitTestNamespace\\", preview_id = \\"some-namespace-id\\" }"
        `);
      });

      it("should create a namespace using configured worker name", async () => {
        writeFileSync("./wrangler.toml", 'name = "otherWorker"', "utf-8");
        mockCreateRequest("otherWorker-UnitTestNamespace");
        const { stdout } = await runWrangler(
          "kv:namespace create UnitTestNamespace"
        );
        expect(stdout).toMatchInlineSnapshot(`
            "ðŸŒ€ Creating namespace with title \\"otherWorker-UnitTestNamespace\\"
            âœ¨ Success!
            Add the following to your configuration file in your kv_namespaces array:
            { binding = \\"UnitTestNamespace\\", id = \\"some-namespace-id\\" }"
            `);
      });

      it("should create a namespace in an environment if configured to do so", async () => {
        mockCreateRequest("worker-customEnv-UnitTestNamespace");
        const { stdout } = await runWrangler(
          "kv:namespace create UnitTestNamespace --env customEnv"
        );
        expect(stdout).toMatchInlineSnapshot(`
          "ðŸŒ€ Creating namespace with title \\"worker-customEnv-UnitTestNamespace\\"
          âœ¨ Success!
          Add the following to your configuration file in your kv_namespaces array under [env.customEnv]:
          { binding = \\"UnitTestNamespace\\", id = \\"some-namespace-id\\" }"
        `);
      });
    });

    describe("list", () => {
      function mockListRequest(namespaces: unknown[]) {
        const requests = { count: 0 };
        setMock(
          "/accounts/:accountId/storage/kv/namespaces\\?:qs",
          ([_url, accountId, query], init) => {
            requests.count++;
            expect(accountId).toEqual("some-account-id");
            expect(query).toContain("per_page=100");
            expect(query).toContain("order=title");
            expect(query).toContain("direction=asc");
            expect(query).toContain("page=");
            expect(init).toBe(undefined);
            const pageSize = Number(/\bper_page=(\d+)\b/.exec(query)[1]);
            const page = Number(/\bpage=(\d+)/.exec(query)[1]);
            return namespaces.slice((page - 1) * pageSize, page * pageSize);
          }
        );
        return requests;
      }

      it("should list namespaces", async () => {
        const KVNamespaces = [
          { title: "title-1", id: "id-1" },
          { title: "title-2", id: "id-2" },
        ];
        mockListRequest(KVNamespaces);
        const { stdout } = await runWrangler("kv:namespace list");
        const namespaces = JSON.parse(stdout);
        expect(namespaces).toEqual(KVNamespaces);
      });

      it("should make multiple requests for paginated results", async () => {
        // Create a lot of mock namespaces, so that the cfetch requests will be paginated
        const KVNamespaces = [];
        for (let i = 0; i < 550; i++) {
          KVNamespaces.push({ title: "title-" + i, id: "id-" + i });
        }
        const requests = mockListRequest(KVNamespaces);
        const { stdout } = await runWrangler("kv:namespace list");
        const namespaces = JSON.parse(stdout);
        expect(namespaces).toEqual(KVNamespaces);
        expect(requests.count).toEqual(6);
      });
    });

    describe("delete", () => {
      function mockDeleteRequest(expectedNamespaceId: string) {
        const requests = { count: 0 };
        setMock(
          "/accounts/:accountId/storage/kv/namespaces/:namespaceId",
          "DELETE",
          ([_url, accountId, namespaceId]) => {
            requests.count++;
            expect(accountId).toEqual("some-account-id");
            expect(namespaceId).toEqual(expectedNamespaceId);
            return null;
          }
        );
        return requests;
      }

      function writeWranglerConfig() {
        writeFileSync(
          "./wrangler.toml",
          [
            'name = "otherWorker"',
            "kv_namespaces = [",
            '  { binding = "someBinding", id = "bound-id", preview_id = "preview-bound-id" }',
            "]",
            "",
            "[env.some-environment]",
            "kv_namespaces = [",
            '  { binding = "someBinding", id = "env-bound-id", preview_id = "preview-env-bound-id" }',
            "]",
          ].join("\n"),
          "utf-8"
        );
      }

      it("should delete a namespace specified by id", async () => {
        const requests = mockDeleteRequest("some-namespace-id");
        await runWrangler(
          `kv:namespace delete --namespace-id some-namespace-id`
        );
        expect(requests.count).toEqual(1);
      });

      it("should delete a namespace specified by binding name", async () => {
        writeWranglerConfig();
        const requests = mockDeleteRequest("bound-id");
        await runWrangler(`kv:namespace delete --binding someBinding`);
        expect(requests.count).toEqual(1);
      });

      it("should delete a preview namespace specified by binding name", async () => {
        writeWranglerConfig();
        const requests = mockDeleteRequest("preview-bound-id");
        await runWrangler(
          `kv:namespace delete --binding someBinding --preview`
        );
        expect(requests.count).toEqual(1);
      });

      it("should error if a given binding name is not in the configured kv namespaces", async () => {
        writeWranglerConfig();
        const { stderr } = await runWrangler(
          `kv:namespace delete --binding otherBinding`
        );
        expect(stderr).toMatchInlineSnapshot(`
          "wrangler kv:namespace delete

          Deletes a given namespace.

          Flags:
            -c, --config   Path to .toml configuration file  [string]
            -h, --help     Show help  [boolean]
            -v, --version  Show version number  [boolean]

          Options:
            -l, --local         Run on my machine  [boolean] [default: false]
                --binding       The name of the namespace to delete  [string]
                --namespace-id  The id of the namespace to delete  [string]
                --env           Perform on a specific environment  [string]
                --preview       Interact with a preview namespace  [boolean]
          Not able to delete namespace.
          A namespace with binding name \\"otherBinding\\" was not found in the configured \\"kv_namespaces\\"."
        `);
      });

      it("should delete a namespace specified by binding name in a given environment", async () => {
        writeWranglerConfig();
        const requests = mockDeleteRequest("env-bound-id");
        await runWrangler(
          `kv:namespace delete --binding someBinding --env some-environment`
        );
        expect(requests.count).toEqual(1);
      });

      it("should delete a preview namespace specified by binding name in a given environment", async () => {
        writeWranglerConfig();
        const requests = mockDeleteRequest("preview-env-bound-id");
        await runWrangler(
          `kv:namespace delete --binding someBinding --env some-environment --preview`
        );
        expect(requests.count).toEqual(1);
      });
    });
  });
});
