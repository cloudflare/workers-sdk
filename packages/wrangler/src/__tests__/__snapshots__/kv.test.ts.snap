// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wrangler > kv namespace > create > wrangler.json > should create a namespace 1`] = `
"ðŸŒ€ Creating namespace with title \\"worker-UnitTestNamespace\\"
âœ¨ Success!
Add the following to your configuration file in your kv_namespaces array:
{
  \\"kv_namespaces\\": [
    {
      \\"binding\\": \\"UnitTestNamespace\\",
      \\"id\\": \\"some-namespace-id\\"
    }
  ]
}"
`;

exports[`wrangler > kv namespace > create > wrangler.json > should create a namespace in an environment if configured to do so 1`] = `
"ðŸŒ€ Creating namespace with title \\"worker-customEnv-UnitTestNamespace\\"
âœ¨ Success!
Add the following to your configuration file in your kv_namespaces array under [env.customEnv]:
{
  \\"kv_namespaces\\": [
    {
      \\"binding\\": \\"UnitTestNamespace\\",
      \\"id\\": \\"some-namespace-id\\"
    }
  ]
}"
`;

exports[`wrangler > kv namespace > create > wrangler.json > should create a namespace using configured worker name 1`] = `
"ðŸŒ€ Creating namespace with title \\"other-worker-UnitTestNamespace\\"
âœ¨ Success!
Add the following to your configuration file in your kv_namespaces array:
{
  \\"kv_namespaces\\": [
    {
      \\"binding\\": \\"UnitTestNamespace\\",
      \\"id\\": \\"some-namespace-id\\"
    }
  ]
}"
`;

exports[`wrangler > kv namespace > create > wrangler.json > should create a preview namespace if configured to do so 1`] = `
"ðŸŒ€ Creating namespace with title \\"worker-UnitTestNamespace_preview\\"
âœ¨ Success!
Add the following to your configuration file in your kv_namespaces array:
{
  \\"kv_namespaces\\": [
    {
      \\"binding\\": \\"UnitTestNamespace\\",
      \\"preview_id\\": \\"some-namespace-id\\"
    }
  ]
}"
`;

exports[`wrangler > kv namespace > create > wrangler.toml > should create a namespace 1`] = `
"ðŸŒ€ Creating namespace with title \\"worker-UnitTestNamespace\\"
âœ¨ Success!
Add the following to your configuration file in your kv_namespaces array:
[[kv_namespaces]]
binding = \\"UnitTestNamespace\\"
id = \\"some-namespace-id\\"
"
`;

exports[`wrangler > kv namespace > create > wrangler.toml > should create a namespace in an environment if configured to do so 1`] = `
"ðŸŒ€ Creating namespace with title \\"worker-customEnv-UnitTestNamespace\\"
âœ¨ Success!
Add the following to your configuration file in your kv_namespaces array under [env.customEnv]:
[[kv_namespaces]]
binding = \\"UnitTestNamespace\\"
id = \\"some-namespace-id\\"
"
`;

exports[`wrangler > kv namespace > create > wrangler.toml > should create a namespace using configured worker name 1`] = `
"ðŸŒ€ Creating namespace with title \\"other-worker-UnitTestNamespace\\"
âœ¨ Success!
Add the following to your configuration file in your kv_namespaces array:
[[kv_namespaces]]
binding = \\"UnitTestNamespace\\"
id = \\"some-namespace-id\\"
"
`;

exports[`wrangler > kv namespace > create > wrangler.toml > should create a preview namespace if configured to do so 1`] = `
"ðŸŒ€ Creating namespace with title \\"worker-UnitTestNamespace_preview\\"
âœ¨ Success!
Add the following to your configuration file in your kv_namespaces array:
[[kv_namespaces]]
binding = \\"UnitTestNamespace\\"
preview_id = \\"some-namespace-id\\"
"
`;
